### ISSUE TITLE
Feature Wishlist

### ISSUE THREAD

**Issue by nicatronTg** [NyxStudios/ModBox#6]
I'm creating a single issue for tracking ideas that are just "ideas" rather than any real planned architecture systems at the moment. If anything becomes real enough that we need another issue, it can be created.

- [ ] Client/server handshake rewrite to authenticate that ModBox is speaking with a ModBox server (to offer upgraded connection protocol)
- [ ] Client/server handshake authentication with Steam
- [ ] Ability to stream custom music, sounds, textures, and other content with a custom format to clients. An example format is the Garry's Mod [addon package](http://wiki.garrysmod.com/page/Workshop_Addon_Creation) that contains a folder structure for models, materials, sounds, and lua scripts. Clients should be able to download these either directly from the server, or from an external HTTP source (ala [sv_downloadurl](https://developer.valvesoftware.com/wiki/Sv_downloadurl)). This is a fairly big one, as it should happen during the connection phase (and thus, needs to happen following an upgraded handshake to verify that ModBox is the server. 
- [ ] Ability to load custom sound, and issue sound playback to clients arbitrarily. Includes music, etc. Probably depends on custom sending content first, and protocol changes.
- [ ] Ability to load custom textures for the purpose of map tiles.
- [ ] Ability to load custom textures for the purpose of new weapons.
- [ ] Ability to load custom textures for the purpose of new NPCs.
- [ ] Custom NPCs.
- [ ] Custom weapons.
- [ ] Custom tiles (new minerals, for instance).
- [ ] A lot of this depends on either protocol changes, or at least some form of client side scripting. My suggestion would be to implement a sandboxed, clientside scripting language, such as either Lua, Ruby, Javascript, CoffeeScript, etc that can be loaded on both the server and the client, obsoleting the existing method of writing C# plugins and loading them via dll. Sandboxing means that we can prevent reaching out of the FS, or limiting things like network libraries to make joining servers with custom code more feasible. I'm pulling more inspiration from Gmod here, as in how it does custom network messaging between scripts. This allows for arbitrary expansion of the protocol by allowing server and client scripts to communicate through strings and other messages. See [here](http://wiki.garrysmod.com/page/Net_Library_Usage).
- [ ] Chatbox improvements. It seems silly that Terraria limits things like length, prevents copy & pasting, and clicking links.
- [ ] Arbitrary UI framework. Can we adopt something like [gwen-dotnet](https://code.google.com/p/gwen-dotnet/) without too much hassle? Is this too big of a project? Maybe. Is it useful? Hell yes.
- [ ] Need a logo.
- [ ] Need a versioning system for releases.
- [ ] Need a style guide.
- [ ] Need a splash screen rewrite too?
- [ ] Need bamboo integration? Possible with mono, not with XNA?
- [ ] Custom game rules. Something that could disable, say, all Terraria vanilla stuff, and only load basic physics and stuff. Yes, lots of work.
- [ ] Master server list, with ability to join servers in game, and remove the need for things like external server browser sites. Show ModBox servers, with player counts, and enable discovery of servers with tags (whitelist required, steam integration, etc).

Misc tasks, unrelated to code:
- [x] Contact Re-Logic about being able to be sold or distributed on Steam. This is complete -- waiting on further contact if this is possible.
- [x] Contact Valve to see about getting the Steam interoperability files fixed from Steamworks. This is complete -- waiting on further contact from Valve (last we heard was that someone was looking into it).

**Comment by nicatronTg** [comment]
Probably the most important and significant thing here is that nearly everything depends on the ability to create and maintain a separate client/server communication channel between ModBox and its associated servers, versus stock clients. We can still improve/offer functionality to vanilla, but the real power here comes from controlling both ends.

**Comment by Olink** [comment]
My thoughts, and yes I know, this is just a list of hypothetical ideas.

* Terraria support paste.  It supports copy if it supported highlighting text.
* Scripting language while probably good, sounds like a lot of work for little gain.  Not to point fingers but a certain someone mentioned not wanting to start from scratch on something and thats what that someone will have to do if we move to a scripting language because that someone is the only one with experience in scripting languages and GMod.
* Custom UI elements is far far far far down the road.
* Would be cool to get Visual Studio and the plugin that loads a style guide file to make this easier.

**Comment by nicatronTg** [comment]
O, didn't know about c & p.

Yep, all of those things are big projects, definitely. Just wanted to write them down somewhere (have been bouncing in my head for a while).

**Comment by Olink** [comment]
That being said it would be cool to swap out the chat box with the one MarioE uses in Raptor.

Also, I don't think we need bamboo for this unless we start writing unit tests for code.
